BrowserServicesKit/Sources/Suggestions/SuggestionProcessing.swift

```diff
@@ -42,263 +50,241 @@ final class SuggestionProcessing {
-                switch ($0.item, $1.item) {
-                // place open tab suggestions on top
-                case (.openTab, .openTab): break
-                case (.openTab, _): return true
-                case (_, .openTab): return false
-                default: break
-                }
-                return $0.score > $1.score
-            }
-            // Create suggestion array
-            .compactMap {
-                switch $0.item {
-                case .bookmark(let bookmark):
-                    switch platform {
-                    case .desktop: return Suggestion(bookmark: bookmark)
-                    case .mobile: return Suggestion(bookmark: bookmark, allowedInTopHits: true)
-                    }
-
-                case .history(let historyEntry):
-                    return Suggestion(historyEntry: historyEntry)
-                case .internalPage(let internalPage):
-                    return Suggestion(internalPage: internalPage)
-                case .openTab(let tab):
-                    return Suggestion(tab: tab)
-                }
-            }
+        // If the suggestion is based solely on an open tab
+        if suggestionKinds == [.browserTab] {
+            // Don't include open tabs in top hits by default
+            return false
+        }
 
-        return result
+        // Other kinds of suggestion can be included in top hits
+        return true
     }
 
-    // MARK: - Top Hits
+    /// Handles special case for open tab suggestions in top hits
+    private func handleTopHitsOpenTabCase(_ topHitsDeduped: some Collection<(suggestion: ScoredSuggestion, kinds: Set<ScoredSuggestion.Kind>)>) -> [ScoredSuggestion] {
+        var result = topHitsDeduped.map(\.suggestion)
 
-    /// Take the top two items from the suggestions, but only up to the first suggestion that is not allowed in top hits
-    private func topHits(from suggestions: [Suggestion]) -> [Suggestion] {
-        var topHits = [Suggestion]()
+        // If the top suggestion is open tab based and also a history entry, bookmark or favorite…
+        guard let topHit = topHitsDeduped.first,
+              topHit.kinds.contains(.browserTab),
+              topHit.kinds.intersects([.historyEntry, .bookmark, .favorite]) else { return result }
 
-        for suggestion in suggestions {
-            guard topHits.count < Self.maximumNumberOfTopHits else { break }
+        // Choose new suggestion kind based on highest quality non-open tab suggestion type
+        let newSuggestionKind = if topHit.suggestion.kind == .browserTab {
+            topHit.kinds.filter { $0 != .browserTab }.max(by: { $0.quality < $1.quality }) ?? .browserTab
+        } else {
+            ScoredSuggestion.Kind.browserTab
+        }
 
-            if suggestion.allowedInTopHits {
-                topHits.append(suggestion)
-            } else {
-                break
-            }
+        // …we split the open tab/other suggestion into separate suggestions…
```

Comment:
The main thing of the project: handling the special case of the Open Tab suggestion getting into the Top Hits.